File: sdk/backd-js/src/jobs.ts
import type { EnqueueOptions } from './types.ts'
import { HttpClient } from './http.ts'

export class JobsModule {
  private http: HttpClient
  private internalUrl: string
  private app: string

  constructor(http: HttpClient, internalUrl: string, app: string) {
    this.http = http
    this.internalUrl = internalUrl
    this.app = app
  }

  async enqueue(fn: string, payload?: unknown, opts?: EnqueueOptions): Promise<{ id: string }> {
    const body: Record<string, unknown> = {
      app: this.app,
      function: fn,
      input: payload ? JSON.stringify(payload) : '{}',
      trigger: 'sdk',
    }

    if (opts?.delay) {
      const now = new Date()
      const ms = parseDuration(opts.delay)
      body.run_at = new Date(now.getTime() + ms).toISOString()
    }

    if (opts?.max_attempts !== undefined) {
      body.max_attempts = opts.max_attempts
    }

    return this.http.request<{ id: string }>({
      method: 'POST',
      baseUrl: this.internalUrl,
      path: '/internal/jobs',
      body,
      noAuth: true,
    })
  }
}

function parseDuration(s: string): number {
  const match = s.match(/^(\d+)(ms|s|m|h|d)$/)
  if (!match) return 0
  const n = parseInt(match[1], 10)
  switch (match[2]) {
    case 'ms': return n
    case 's': return n * 1000
    case 'm': return n * 60_000
    case 'h': return n * 3_600_000
    case 'd': return n * 86_400_000
    default: return 0
  }
}

File: sdk/backd-go/jobs.go
package backd

import (
	"context"
	"encoding/json"
	"time"
)

// JobsClient handles background job operations (server-side only).
type JobsClient struct {
	http        *httpClient
	internalURL string
	app         string
}

func newJobsClient(h *httpClient, internalURL, app string) *JobsClient {
	return &JobsClient{http: h, internalURL: internalURL, app: app}
}

// Enqueue submits a background job for processing.
func (j *JobsClient) Enqueue(ctx context.Context, fn string, payload any, opts ...EnqueueOptions) (string, error) {
	body := map[string]any{
		"app":      j.app,
		"function": fn,
		"input":    "{}",
		"trigger":  "sdk",
	}

	if payload != nil {
		data, err := json.Marshal(payload)
		if err != nil {
			return "", err
		}
		body["input"] = string(data)
	}

	if len(opts) > 0 {
		o := opts[0]
		if o.Delay != "" {
			ms := parseDuration(o.Delay)
			body["run_at"] = time.Now().Add(ms).UTC().Format(time.RFC3339)
		}
		if o.MaxAttempts > 0 {
			body["max_attempts"] = o.MaxAttempts
		}
	}

	result, err := request[map[string]string](ctx, j.http, requestOptions{
		method:  "POST",
		baseURL: j.internalURL,
		path:    "/internal/jobs",
		body:    body,
		noAuth:  true,
	})
	if err != nil {
		return "", err
	}
	return result["job_id"], nil
}

func parseDuration(s string) time.Duration {
	if len(s) < 2 {
		return 0
	}
	unit := s[len(s)-1]
	numStr := s[:len(s)-1]
	// Handle "ms" suffix
	if len(s) >= 3 && s[len(s)-2:] == "ms" {
		numStr = s[:len(s)-2]
		unit = 'M' // sentinel
	}

	var n int
	for _, c := range numStr {
		if c >= '0' && c <= '9' {
			n = n*10 + int(c-'0')
		}
	}

	switch unit {
	case 'M': // ms sentinel
		return time.Duration(n) * time.Millisecond
	case 's':
		return time.Duration(n) * time.Second
	case 'm':
		return time.Duration(n) * time.Minute
	case 'h':
		return time.Duration(n) * time.Hour
	case 'd':
		return time.Duration(n) * 24 * time.Hour
	default:
		return 0
	}
}
